sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamingQueryCache.scala

```diff
@@ -228,62 +212,109 @@ private[connect] class SparkConnectStreamingQueryCache(
+        case Some(_) => // Inactive query waiting for expiration. Do nothing.
+          logInfo(
+            log"Waiting for the expiration for id: ${MDC(QUERY_ID, id)} " +
+              log"runId: ${MDC(QUERY_RUN_ID, runId)} in " +
+              log"session ${MDC(SESSION_ID, v.sessionId)}")
+
+        case None => // Active query, check if it is stopped. Enable timeout if it is stopped.
+          val isActive = v.query.isActive && Option(v.session.streams.get(id)).nonEmpty
 
-          case Some(_) => // Inactive query waiting for expiration. Do nothing.
+          if (!isActive) {
             logInfo(
-              log"Waiting for the expiration for id: ${MDC(QUERY_ID, id)} " +
+              log"Marking query id: ${MDC(QUERY_ID, id)} " +
                 log"runId: ${MDC(QUERY_RUN_ID, runId)} in " +
-                log"session ${MDC(SESSION_ID, v.sessionId)}")
-
-          case None => // Active query, check if it is stopped. Enable timeout if it is stopped.
-            val isActive = v.query.isActive && Option(v.session.streams.get(id)).nonEmpty
-
-            if (!isActive) {
-              logInfo(
-                log"Marking query id: ${MDC(QUERY_ID, id)} " +
-                  log"runId: ${MDC(QUERY_RUN_ID, runId)} in " +
-                  log"session ${MDC(SESSION_ID, v.sessionId)} inactive.")
-              val expiresAtMs = nowMs + stoppedQueryInactivityTimeout.toMillis
-              queryCache.put(k, v.copy(expiresAtMs = Some(expiresAtMs)))
-              // To consider: Clean up any runner registered for this query with the session holder
-              // for this session. Useful in case listener events are delayed (such delays are
-              // seen in practice, especially when users have heavy processing inside listeners).
-              // Currently such workers would be cleaned up when the connect session expires.
-            }
-        }
+                log"session ${MDC(SESSION_ID, v.sessionId)} inactive.")
+            val expiresAtMs = nowMs + stoppedQueryInactivityTimeout.toMillis
+            queryCache.put(k, v.copy(expiresAtMs = Some(expiresAtMs)))
+            // To consider: Clean up any runner registered for this query with the session holder
+            // for this session. Useful in case listener events are delayed (such delays are
+            // seen in practice, especially when users have heavy processing inside listeners).
+            // Currently such workers would be cleaned up when the connect session expires.
+          }
       }
+    })
 
-      taggedQueries.toArray.foreach { case (key, value) =>
-        value.zipWithIndex.toArray.foreach { case (queryKey, i) =>
-          if (queryCache.contains(queryKey)) {
-            value.remove(i)
-          }
+    // Removes any tagged queries that do not correspond to cached queries.
+    taggedQueries.forEach((key, value) => {
+      if (value.filter(k => queryCache.containsKey(k))) {
+        taggedQueries.remove(key, value)
```

Comment:
Hm, it has to remove the element within `value` instead of the whole entry IIRC.

Reply:
You're right, and the code actually does that, though it's not quite obvious; `value.filter` will return true only when `value` is empty, so semantically the piece of code means,
1. Remove elements that do not satisfy `queryCache.containsKey(k)` from `value`.
2. If `value` is empty after filtering those elements, `value` will be removed from `taggedQueries`.

Reply:
ah okie, then lgtm